crates/release_plz_core/src/git/github_graphql.rs

```diff
@@ -0,0 +1,168 @@
+use std::{collections::HashMap, path::Path};
+
+use anyhow::Result;
+use base64::prelude::*;
+use git_cmd::Repo;
+use serde_json::Value;
+use tokio::fs;
+use tracing::{debug, trace};
+use url::Url;
+
+use crate::git::backend::Remote;
+use crate::GitClient;
+
+fn get_graphql_endpoint(remote: &Remote) -> Url {
+    let mut base_url = remote.base_url.clone();
+    base_url.set_path("graphql");
+
+    base_url
+}
+
+/// Commit all the changes (except typestates) that are present in the repository
+/// using GitHub's [GraphQL api](https://docs.github.com/en/graphql/reference/mutations#createcommitonbranch).
+pub async fn commit_changes(
+    client: &GitClient,
+    repo: &Repo,
+    message: &str,
+    branch: &str,
+) -> Result<()> {
+    let current_head = repo.current_commit_hash()?;
+    let deletions = removed_files(repo)?;
+    let changes = changed_files(repo)?;
+
+    let commit_query = format_commit_query(
+        &client.remote.owner_slash_repo(),
+        branch,
+        message,
+        &current_head,
+        &deletions,
+        &changes,
+        repo.directory(),
+    )
+    .await?;
+
+    let graphql_endpoint = get_graphql_endpoint(&client.remote);
+
+    debug!("Sending createCommitOnBranch to {}", graphql_endpoint);
+    trace!("{}", commit_query);
+
+    let json_body: HashMap<&str, &str> = [("query", commit_query.as_str())].into();
+
+    let res: Value = client
+        .client
+        .post(graphql_endpoint)
+        .json(&json_body)
+        .send()
+        .await?
+        .json()
+        .await?;
+
+    if let Some(errors) = res.get("errors").and_then(Value::as_array) {
+        anyhow::bail!(
+            "createCommitOnBranch returned errors: {:?}",
+            serde_json::to_string(errors)?
+        );
+    }
+
+    Ok(())
+}
+
+// get the list of changes in repository excluding typechanges and removed files
+fn changed_files(repo: &Repo) -> Result<Vec<String>> {
+    repo.changes(|line| !line.starts_with("T ") && !line.starts_with("D "))
+}
+
+// get the list of removed files in repository
+fn removed_files(repo: &Repo) -> Result<Vec<String>> {
+    repo.changes(|line| line.starts_with("D "))
+}
+
+// format a graphql query to create commit on branch
+async fn format_commit_query(
+    owner_and_repo: &str,
+    branch: &str,
+    message: &str,
+    current_head: &str,
+    deletions: &[impl AsRef<Path>],
+    additions: &[impl AsRef<Path>],
+    repo_dir: impl AsRef<Path>,
```

Comment:
can you collect all these parameters into a struct? E.g. `GitHubCommit`

Reply:
done